pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+			expectedImage := "apache"
+
+			BeforeEach(func() {
+				initObjs = append(initObjs, fedResource)
+				resource = newPod("test-pod", expectedImage)
+			})
+
+			It("should update the resource", func() {
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+				fedPodObj, _, err := getTemplateField(fedPod)
+				Expect(err).ToNot(HaveOccurred())
+				expectObj, _, err := getTemplateField(fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(fedPodObj["spec"]).ToNot(BeEquivalentTo(expectObj["spec"]))
+
+				mapObj, _, _ := unstructured.NestedSlice(
+					fedPod.Object,
+					ctlutil.SpecField, ctlutil.TemplateField, "spec", "containers",
+				)
+				image := mapObj[0].(map[string]interface{})["image"]
+				Expect(image).To(BeEquivalentTo(expectedImage))
+			})
+		})
+
+		When("on update the Kube API returns an error that is not a NotFound", func() {
+			BeforeEach(func() {
+				clientPatcher = &fakeClientWithUpdateError{"test error"}
+			})
+
+			It("should return an error and not create the resource", func() {
+				err := f.Distribute(resource, clusterNames...)
+				Expect(err).To(MatchError(HavePrefix("test error")))
+			})
+		})
+
+		When("the resource is not in the Kube API", func() {
+			It("should create the resource", func() {
+				// Sanity check
+				_, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).To(HaveOccurred())
+
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(fedPod).ToNot(BeNil())
+				Expect(fedPod).To(BeEquivalentTo(fedResource))
+			})
+		})
+
+		When("the resource needs to be distributed to specific clusters", func() {
+			BeforeEach(func() {
+				clusterNames = []string{"cluster1", "cluster2"}
+			})
+
+			It("should set the placement to be an explicit list of clusters", func() {
+				// This test implicitly tests that the nested hierarchy of fields exists
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				mapObj, found, err := getClustersField(fedPod)
+				fedClusterNames := unpackClusterNames(mapObj)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+				Expect(fedClusterNames).To(BeEquivalentTo(clusterNames))
+			})
+
+			It("the returned resource's placement should not contain an empty matcher", func() {
+				// This test implicitly tests that the nested hierarchy of fields exists
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				mapObj, found, err := getMatchLabelsField(fedPod)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).ToNot(BeTrue(), "Hierarchy of fields is present!")
```

Comment:
Can we change this to _"The matchLabels field is missing!"_? Also at the other call sites.  

We could  go further and have `getMatchLabelsField` do the checking that's on lines 290-291 and have it return only a valid `mapObj` - that would simplify the call sites. Same with `getClustersField` and `getTemplateField`.

Reply:
> Can we change this to _"The matchLabels field is missing!"_? Also at the other call sites.

Same as other sites. 

> We could go further and have `getMatchLabelsField` do the checking that's on lines 290-291 and have it return only a valid `mapObj` - that would simplify the call sites. Same with `getClustersField` and `getTemplateField`.

There are variations of the assertions, it's not always the same thing that is expected.